### DISCUSSION TITLE
Shared workspaces across monitors

### DISCUSSION THREAD

**Discussion by YannikSc** [YaLTeR/niri#446] [Ideas]
As mentioned niri is inspired by PaperWM. Which I absolutely love, btw! One feature of PaperWM however, which I heavily used, is missing in my opinion and that one is, that the workspaces are shared across the monitors.

While it looks a bit messy, code wise, to implement that I don't think it's too far off and it would be a super cool feature, or am I the only one feeling that?

**Comment by YaLTeR** [comment]
You can move workspaces to a different monitor with a set of binds like:
```
    Mod+Shift+Alt+Left  { move-workspace-to-monitor-left; }
    Mod+Shift+Alt+Down  { move-workspace-to-monitor-down; }
    Mod+Shift+Alt+Up    { move-workspace-to-monitor-up; }
    Mod+Shift+Alt+Right { move-workspace-to-monitor-right; }
```

That's as shared as they will get. I personally find the way Paper does shared workspaces a bit weird, especially when your monitors have different resolution.

**Reply by salman-farooq-sh** [reply]
Is there a wayland solution for treating multiple monitors as a single large one like X11 had?

**Reply by YaLTeR** [reply]
Wayland has no say on this. The compositor is free to do this, but it needs to be implemented.

**Comment by ArijanJ** [comment]
I would also like to see this happen, but since you've mentioned not liking it, here are a few of my use cases for it:
- watching a movie on workspace 9 on secondary monitor, being able to bring it to my main monitor by doing mod+ctrl+9 while focused on my main one - if I do this right now, I have to move to an empty workspace on my main monitor, then focus the secondary one and move the workspace to the left output
- having a "swap monitor workspaces" bind for something similar to the above
- having an app hidden somewhere out of view and instantly being able to bring it to whichever monitor i want by doing mod+ctrl+3 (and then alt+tab when I'm done with it)

The main issue is that all workspaces are locked to a monitor, even named ones. Do you think something like this could ever be done, or is it simply not in scope for niri? I guess what I'm trying to do is use the horizontal scrolling but have more fine-grained control over where each workspace goes.

**Reply by YaLTeR** [reply]
> having an app hidden somewhere out of view and instantly being able to bring it to whichever monitor i want

Sounds like #845 will solve this

> The main issue is that all workspaces are locked to a monitor, even named ones.

Btw there's a `move-workspace-to-monitor` command that might help in some of these cases

**Comment by noctuid** [comment]
The current issues I had when trying to share workspaces across windows by using named workspaces only:

- You can kind of configure waybar to handle this with `"all-outputs": true` but there's no way to control/get a consistent workspace order, which is confusing
- I couldn't write a script that will switch to any workspace on the current monitor

@YaLTeR  Do you know why this doesn't work?
```sh
current_output=$(niri msg --json focused-output | jq '.name')
workspace_output=$(niri msg --json workspaces \
                       | jq ".[] | select(.name == \"$1\").output")

if [[ $current_output != "$workspace_output" ]]; then
    # TODO there is no focus-monitor?
    while true; do
        niri msg action focus-monitor-next
        new_output=$(niri msg --json focused-output | jq '.name')
        if [[ $new_output == "$workspace_output" ]]; then
            break
        fi
    done
fi

niri msg action focus-workspace "$1"

if [[ $current_output != "$workspace_output" ]]; then
    focused_workspace=$(niri msg --json workspaces \
        | jq ".[] | select(.is_focused == true).name")
    echo "Focused: $focused_workspace, move to: $current_output"
    niri msg action move-workspace-to-monitor "$current_output"
fi
```

I tried a couple approaches, but none of them worked. `move-workspace-to-monitor` works if I call it in a terminal on the workspace, but if I call it in this script, it doesn't.  I thought maybe it's because the other actions hadn't completed.  I tried adding sleep, but even when that echo outputs the expected information, it does not move the workspace. It might be nice if there was an action that specified both the workspace and the monitor.

I like the way herbstluftwm handles things because it supports both methods.  You can optionally lock workspaces to specific monitors so that switching to them will switch to that monitor if on another monitor, but otherwise you can switch to any workspace on any monitor.

**Comment by ArijanJ** [comment]
@YaLTeR would it be possible to have one set of global workspaces that can be accessed from any monitor? (up/down movement)

Ideally, it would skip the ones shown on the other monitors. I would be happy to try and PR this if you think it's plausible with the current code base.

**Reply by YaLTeR** [reply]
No, that is not how niri works or how I would like it to work. It also doesn't make much sense spatially. You also cannot easily refactor niri to work this way.

**Reply by jonaspleyer** [reply]
Would there be a way to define a shortcut which automatically moves all workspaces down. I.e. redefine `Mod+U` such that it affects all monitors simultanously.

**Reply by Sempyos** [reply]
No, you would need to try to script it via the IPC

**Reply by jonaspleyer** [reply]
Do you think that this might be something that people would consider a valid workaround for multi-monitor workspaces? If so then I might spend some time to develop this feature. Would this even be accepted by niri?
